streaming/base/util.py

```diff
@@ -196,3 +204,138 @@ def get_import_exception_message(package_name: str, extra_deps: str) -> str:
+            defaults to 60, same as streaming.download_file
+
+    Returns:
+        int: count of files downloaded during function call
+    """
+    # Import here to avoid circular import error
+    from streaming.base.storage.download import download_file
+    from streaming.base.storage.upload import CloudUploader
+
+    if not folder_urls:
+        logger.warning('No partitions exist, no index merged')
+        return 0
+
+    # This is the index json file name, e.g., it is index.json as of 0.6.0
+    index_basename = get_index_basename()
+
+    cu = CloudUploader.get(out, keep_local=True, exist_ok=True)
+    if os.path.exists(os.path.join(cu.local, index_basename)) and overwrite:
+        logger.warning('Merged index already exists locally. no index merged if overwrite=False')
+        return 0
+
+    # Remove '/' from right, so os.path.basename gives relative path to each folder
+    urls = []
+    for url in folder_urls:
+        if type(url) is str:
+            urls.append(url.rstrip('/').strip())
+        else:
+            urls.append((url[0].rstrip('/').strip(), url[1].rstrip('/').strip()))
+
+    # Determine if we need to call download_file.
+    download = False
+    for url in urls:
+        if type(url) is tuple:
+            # If driver cannot access the local path, download = True
+            download = not os.path.exists(url[0])
+        else:
+            # If url is a remote, download = True, False otherwise
+            download = urllib.parse.urlparse(url).scheme != ''
+
+        # As long as one index file needs download, we download them all to keep it simple
+        if download:
+            break
+
+    print('download = ', download)
+    # Prepare a temp folder to download index.json rom remote if necessary. Removed in the end.
+    with tempfile.TemporaryDirectory() as temp_root:
+
+        # container for absolute local folder path
+        partitions = []
+        n_downloads = 0
+        for url in urls:
+            if type(url) is tuple:
+                local, remote = url
+            else:
+                local = remote = url
+
+            if download:
+                # If download is needed, download url from remote to temp_root
+                path = urllib.parse.urlparse(remote).path
+                local = os.path.join(temp_root, path.lstrip('/'))
+                try:
+                    remote_url = os.path.join(remote, index_basename)
+                    local_path = os.path.join(local, index_basename)
+                    download_file(remote_url, local_path, download_timeout)
+                    n_downloads += 1
+                except Exception as ex:
+                    raise RuntimeError(f'failed to download index.json {url}') from ex
+
+            if not (os.path.exists(local)):
+                raise FileNotFoundError(
+                    'Folder {local} does not exit or cannot be acceessed by the current process')
```

Comment:
`f` string is missing. `f'Folder {local} does not exit or cannot be acceessed by the current process'`